gax/src/main/java/com/google/api/gax/retrying/ExponentialRetryAlgorithm.java

```diff
@@ -135,9 +135,21 @@ public boolean shouldRetry(TimedAttemptSettings nextAttemptSettings) {
             - nextAttemptSettings.getFirstAttemptStartTimeNanos()
             + nextAttemptSettings.getRandomizedRetryDelay().toNanos();
 
-    return totalTimeSpentNanos <= globalSettings.getTotalTimeout().toNanos()
-        && (globalSettings.getMaxAttempts() <= 0
-            || nextAttemptSettings.getAttemptCount() < globalSettings.getMaxAttempts());
+    int maxAttempts = globalSettings.getMaxAttempts();
+    long totalTimeout = globalSettings.getTotalTimeout().toNanos();
+
+    // If totalTimeout limit is defined, check that it hasn't been crossed
+    if (totalTimeout > 0 && totalTimeSpentNanos > totalTimeout) {
```

Comment:
It looks like if both `totalTimeout <= 0` and `maxAttempts == 0` this will retry forever. Forever retrying seems dangerous.

Reply:
Should users be prevented from retrying however? If so, I think we should add a Precondition to RetrySettings that requires either maxAttempts > 0 or totalTimeout > 0. I think that 
RetrySettings.newBuilder().setMaxAttempts(5).build() should be a valid use case

Reply:
>I think we should add a Precondition to RetrySettings that requires either maxAttempts > 0 or totalTimeout > 0. 

For this, we need to agree upon a default `maxAttempts` or `totalTimeout` other than 0, Also It could conflict with the user's provided configuration where they have left RetrySettings with default values.
https://github.com/googleapis/gax-java/blob/8d7aafcd012d83b85f8733b6ef589fcbbf2ee261/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java#L137-L148

For the above reason, I hope limiting the retries to 0 [here](https://github.com/googleapis/gax-java/pull/835/files#diff-5b59e6b0eb11dc7fcfd76ec7e23e9a6fR138-R140) would be ok.

Reply:
@igorbernstein2 @vam-google Please let me know if this is acceptable.

Reply:
Resolving this in favor of the discussion below